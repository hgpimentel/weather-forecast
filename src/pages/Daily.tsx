import { useState, useContext, useEffect } from "react";
import styled from "styled-components";
import { CenteredContainer, ColumnContainer } from "../components";
import getDailyForecast from "../api/getDailyForecast";
import LocationContext from "../store/location-context";

const DailyContainer = styled(ColumnContainer)`
  margin: 0 20%;
  padding: 2rem;
  border: 5px solid teal;
  border-radius: 1rem;
`;

const CustomCenteredContainer = styled(CenteredContainer)`
  align-items: center;
  justify-content: space-evenly;
  padding: 1rem;
  border-bottom: 2px solid teal;

  & > * {
    flex-grow: 1;
    max-width: 100px;
    display: flex;
    justify-content: center;
  }
`;

interface DailyForecast {
  day: number;
  avgTemp: number;
  sky: number;
  rain: string;
  wind: number;
}

const Daily: React.FC = () => {
  const [dailyForecasts, setDailyForecasts] =
    useState<DailyForecast[] | null>(null);
  const { currentLocation } = useContext(LocationContext);
  const [isLoading, setIsLoading] = useState(false);

  useEffect(() => {
    if (currentLocation) {
      setIsLoading(true);
      getDailyForecast(currentLocation.lat, currentLocation.long)
        .then((forecasts) => {
          if (forecasts) {
            const dailyForecasts = forecasts.map<DailyForecast>((frc) => ({
              day: frc.dt,
              avgTemp: frc.temp.max,
              sky: frc.clouds,
              rain: `${frc.rain ? frc.rain : "No rain"}`,
              wind: frc.wind_speed,
            }));

            setDailyForecasts(dailyForecasts);
            return;
          }
          setDailyForecasts(null);
        })
        .finally(() => {
          setIsLoading(false);
        });
    }
  }, [currentLocation]);

  if (isLoading) return null;

  if (!dailyForecasts)
    return (
      <div>
        <h1>No daily forecasts found!</h1>
      </div>
    );

  return (
    <DailyContainer>
      <CustomCenteredContainer>
        <div>
          <b>Day</b>
        </div>
        <div>
          <b>Avg Temp</b>
        </div>
        <div>
          <b>Sky</b>
        </div>
        <div>
          <b>Rain</b>
        </div>
        <div>
          <b>Wind</b>
        </div>
      </CustomCenteredContainer>
      {dailyForecasts.map((df) => {
        return (
          <CustomCenteredContainer key={df.day}>
            <div>{df.day}</div>
            <div>{df.avgTemp}</div>
            <div>{df.sky}</div>
            <div>{df.rain}</div>
            <div>{df.wind}</div>
          </CustomCenteredContainer>
        );
      })}
    </DailyContainer>
  );
};

export default Daily;
