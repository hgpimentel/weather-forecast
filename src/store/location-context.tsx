import React, { useEffect, useState } from "react";

interface Location {
  city: string;
  country: string;
  lat: number;
  long: number;
}

type LocationType = Location | null;

const DefaultLocation: LocationType = {
  country: "Portugal",
  city: "Lisbon",
  lat: 38.7259284,
  long: -9.137382,
};

const LocationContext = React.createContext<{
  currentLocation: LocationType;
  onLocationChange: (location: LocationType) => void;
}>({
  currentLocation: null,
  onLocationChange: () => {},
});

export const LocationContextProvider: React.FC = ({ children }) => {
  const [location, setLocation] = useState<LocationType>(null);

  useEffect(() => {
    setLocation(DefaultLocation);
  }, []);

  return (
    <LocationContext.Provider
      value={{ currentLocation: location, onLocationChange: setLocation }}
    >
      {children}
    </LocationContext.Provider>
  );
};

export default LocationContext;
