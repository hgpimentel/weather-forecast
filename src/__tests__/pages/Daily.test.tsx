import renderer from "react-test-renderer";
import Daily from "../../pages/Daily";
import { render, screen } from "@testing-library/react";
import axios from "axios";
import { LocationContextProvider } from "../../store/location-context";
import { act } from "react-dom/test-utils";

jest.mock("axios");

describe("Daily page", () => {
  it("renders correctly", () => {
    const component = renderer.create(<Daily />).toJSON();

    expect(component).toMatchSnapshot();
  });

  it("has elements", async () => {
    const mockedAxios = axios as jest.Mocked<typeof axios>;

    mockedAxios.get.mockImplementationOnce(() =>
      Promise.resolve({
        data: {
          lat: 33.44,
          lon: -94.04,
          timezone: "America/Chicago",
          timezone_offset: -18000,
          daily: [
            {
              dt: 1621792800,
              sunrise: 1621768260,
              sunset: 1621818912,
              moonrise: 1621808040,
              moonset: 1621761480,
              moon_phase: 0.39,
              temp: {
                day: 27.79,
                min: 19.03,
                max: 28.16,
                night: 19.63,
                eve: 22.96,
                morn: 19.08,
              },
              feels_like: {
                day: 28.39,
                night: 19.69,
                eve: 23.3,
                morn: 19.52,
              },
              pressure: 1022,
              humidity: 52,
              dew_point: 17.04,
              wind_speed: 3.58,
              wind_deg: 136,
              wind_gust: 7.44,
              weather: [
                {
                  id: 801,
                  main: "Clouds",
                  description: "few clouds",
                  icon: "02d",
                },
              ],
              clouds: 20,
              pop: 0.04,
              uvi: 10.17,
            },
          ],
        },
      })
    );

    await act(async () => {
      render(
        <LocationContextProvider>
          <Daily />
        </LocationContextProvider>
      );
    });

    expect(screen.getByText("Day")).not.toBeNull();

    expect(screen.getByText("Temperature")).not.toBeNull();
  });
});
