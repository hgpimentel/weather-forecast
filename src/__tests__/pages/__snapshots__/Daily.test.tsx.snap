// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Daily page renders correctly 1`] = `null`;
